### ISSUE TITLE
Bleadperl v5.19.6-171-g437e3a7 breaks ROBIN/Want-0.21.tar.gz

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13470]
Migrated from [rt.perl.org#120792](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=120792) (status was 'resolved')

Searchable as RT120792$

**Comment by p5pRT** [comment]
### From @andk

git bisect


---
commit 437e3a7dac994ebace1195549170c81f474d9c20
Author&#8203;: Matthew Horsfall \<WolfSage@&#8203;gmail\.com>
Date&#8203;:   Wed Dec 11 18&#8203;:28&#8203;:21 2013 \-0500

&nbsp;    Optimise out PUSHMARK/RETURN if return is the last statement in a sub\.

diagnostics


---
&nbsp;  Test Summary Report
&nbsp;  \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
&nbsp;  t/assign\.t  \(Wstat&#8203;: 65280 Tests&#8203;: 6 Failed&#8203;: 0\)
&nbsp;    Non\-zero exit status&#8203;: 255
&nbsp;    Parse errors&#8203;: Bad plan\.  You planned 10 tests but ran 6\.
&nbsp;  Files=8\, Tests=142\,  1 wallclock secs \( 0\.04 usr  0\.01 sys \+  0\.08 cusr  0\.02 csys =  0\.15 CPU\)

perl \-V


---
Summary of my perl5 \(revision 5 version 19 subversion 7\) configuration&#8203;:
&nbsp;  Commit id&#8203;: 437e3a7dac994ebace1195549170c81f474d9c20
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=3\.11\-2\-amd64\, archname=x86\_64\-linux
&nbsp;    uname='linux k83 3\.11\-2\-amd64 \#1 smp debian 3\.11\.10\-1 \(2013\-12\-04\) x86\_64 gnulinux '
&nbsp;    config\_args='\-Dprefix=/home/sand/src/perl/repoperls/installed\-perls/perl/v5\.19\.6\-171\-g437e3a7/165a \-Dmyhostname=k83 \-Dinstallusrbinperl=n \-Uversiononly \-Dusedevel \-des \-Ui\_db \-Uuseithreads \-Uuselongdouble \-DDEBUGGING=\-g'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=undef\, usemultiplicity=undef
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2 \-g'\,
&nbsp;    cppflags='\-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.8\.2'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8  
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-lgdbm \-ldb \-ldl \-lm \-lcrypt \-lutil \-lc \-lgdbm\_compat
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    libc=\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.17'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-O2 \-g \-L/usr/local/lib \-fstack\-protector'


Characteristics of this binary \(from libperl\)&#8203;:
&nbsp;  Compile\-time options&#8203;: HAS\_TIMES PERLIO\_LAYERS PERL\_DONT\_CREATE\_GVSV
&nbsp;                        PERL\_HASH\_FUNC\_ONE\_AT\_A\_TIME\_HARD PERL\_MALLOC\_WRAP
&nbsp;                        PERL\_NEW\_COPY\_ON\_WRITE PERL\_PRESERVE\_IVUV
&nbsp;                        PERL\_USE\_DEVEL USE\_64\_BIT\_ALL USE\_64\_BIT\_INT
&nbsp;                        USE\_LARGE\_FILES USE\_LOCALE USE\_LOCALE\_COLLATE
&nbsp;                        USE\_LOCALE\_CTYPE USE\_LOCALE\_NUMERIC USE\_PERLIO
&nbsp;                        USE\_PERL\_ATOF
&nbsp;  Built under linux
&nbsp;  Compiled at Dec 13 2013 18&#8203;:53&#8203;:48
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /home/sand/src/perl/repoperls/installed\-perls/perl/v5\.19\.6\-171\-g437e3a7/165a/lib/site\_perl/5\.19\.7/x86\_64\-linux
&nbsp;    /home/sand/src/perl/repoperls/installed\-perls/perl/v5\.19\.6\-171\-g437e3a7/165a/lib/site\_perl/5\.19\.7
&nbsp;    /home/sand/src/perl/repoperls/installed\-perls/perl/v5\.19\.6\-171\-g437e3a7/165a/lib/5\.19\.7/x86\_64\-linux
&nbsp;    /home/sand/src/perl/repoperls/installed\-perls/perl/v5\.19\.6\-171\-g437e3a7/165a/lib/5\.19\.7
&nbsp;    \.

\-\- 
andreas

**Comment by p5pRT** [comment]
### From @wolfsage

On Sat\, Dec 14\, 2013 at 11&#8203;:22 PM\, Andreas J\. Koenig via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>   t/assign\.t  \(Wstat&#8203;: 65280 Tests&#8203;: 6 Failed&#8203;: 0\)
>     Non\-zero exit status&#8203;: 255
>     Parse errors&#8203;: Bad plan\.  You planned 10 tests but ran 6\.
>   Files=8\, Tests=142\,  1 wallclock secs \( 0\.04 usr  0\.01 sys \+  0\.08 cusr  0\.02 csys =  0\.15 CPU\)

Some background&#8203;:

&nbsp;  $ \./perl \-Ilib assign\.t
&nbsp;  1\.\.10
&nbsp;  ok 1
&nbsp;  ok 2
&nbsp;  ok 3
&nbsp;  ok 4
&nbsp;  ok 5
&nbsp;  ok 6
&nbsp;  Died at assign\.t line 42\.

assign\.t&#8203;:

&nbsp;  32&#8203;: sub backstr :lvalue \{
&nbsp;  33&#8203;:     if \(want\('LVALUE'\)\) \{
&nbsp;  34&#8203;:         carp\("Not in ASSIGN context"\) unless want\('ASSIGN'\);
&nbsp;  35&#8203;:         my $a = want\('ASSIGN'\);
&nbsp;  36&#8203;:         $\_\[0\] = reverse $a;
&nbsp;  37&#8203;:         lnoreturn;
&nbsp;  38&#8203;:     \}
&nbsp;  39&#8203;:     else \{
&nbsp;  40&#8203;:         rreturn scalar reverse $\_\[0\];
&nbsp;  41&#8203;:     \}
&nbsp;  42&#8203;:     die; return;
&nbsp;  43&#8203;:  \}


lnoreturn / rreturn from Want\.pm&#8203;:

&nbsp;  sub rreturn \(@&#8203;\) \{
&nbsp;      if \(want\_lvalue\(1\)\) \{
&nbsp;          croak "Can't rreturn in lvalue context";
&nbsp;      \}
&nbsp;      double\_return\(\);
&nbsp;      return wantarray ? @&#8203;\_ : $\_\[$\#\_\];
&nbsp;  \}

&nbsp;  sub lnoreturn \(\) \{
&nbsp;      if \(\!want\_lvalue\(1\) || \!want\_assign\(1\)\) \{
&nbsp;          croak "Can't lnoreturn except in ASSIGN context";
&nbsp;      \}
&nbsp;      double\_return\(\);
&nbsp;      return disarm\_temp\(my $undef\);
&nbsp;  \}


double\_return from Want\.xs&#8203;:

&nbsp;  void
&nbsp;  double\_return\(\)
&nbsp;    PREINIT&#8203;:
&nbsp;      PERL\_CONTEXT \*ourcx\, \*cx;
&nbsp;    PPCODE&#8203;:
&nbsp;      ourcx = upcontext\(aTHX\_ 0\);
&nbsp;      cx    = upcontext\(aTHX\_ 1\);
&nbsp;      if \(\!cx\)
&nbsp;          Perl\_croak\(aTHX\_ "Can't return outside a subroutine"\);

&nbsp;      ourcx\->cx\_type = CXt\_NULL;
&nbsp;      CvDEPTH\(ourcx\->blk\_sub\.cv\)\-\-;
&nbsp;  \#if HAS\_RETSTACK
&nbsp;      if \(PL\_retstack\_ix > 0\)
&nbsp;          \-\-PL\_retstack\_ix;
&nbsp;  \#endif

&nbsp;      return;

\(My knowledge in this area is very limited\, apologies if this makes no sense\)

What it looks like to me is happening is double\_return pops a stack
frame\, effectively putting the return calls in the sub above lnoreturn
and rreturn\.

My optimisation unfortunately removes the returns \(since they're at
the end of the lnoreturn / rreturn subs\) and so the test gets treated
like this&#8203;:

&nbsp;  32&#8203;: sub backstr :lvalue \{
&nbsp;  33&#8203;:     if \(want\('LVALUE'\)\) \{
&nbsp;  34&#8203;:         carp\("Not in ASSIGN context"\) unless want\('ASSIGN'\);
&nbsp;  35&#8203;:         my $a = want\('ASSIGN'\);
&nbsp;  36&#8203;:         $\_\[0\] = reverse $a;
&nbsp;  37&#8203;:         undef; \# Falls through to die now
&nbsp;  38&#8203;:     \}
&nbsp;  39&#8203;:     else \{
&nbsp;  40&#8203;:         scalar reverse $\_\[0\]; \# Falls through to die now
&nbsp;  41&#8203;:     \}
&nbsp;  42&#8203;:     die; return;
&nbsp;  43&#8203;:  \}

What I don't know is if Want is violating some rules with regards to
the Perl internals and doing something it shouldn't\.\.\.

This patch does however fix the issue since the extra scope prevents
the optimisation from working\.  \(That is\, until we some day make this
optimisation smarter\)&#8203;:

&nbsp;  \-\-\- Want\.pm   2012\-02\-29 13&#8203;:54&#8203;:14\.000000000 \-0500
&nbsp;  \+\+\+ Want\.pm 2013\-12\-15 08&#8203;:43&#8203;:55\.000000000 \-0500
&nbsp;  @&#8203;@&#8203; \-169\,7 \+169\,7 @&#8203;@&#8203;
&nbsp;           croak "Can't rreturn in lvalue context";
&nbsp;       \}
&nbsp;       double\_return\(\);
&nbsp;  \-    return wantarray ? @&#8203;\_ : $\_\[$\#\_\];
&nbsp;  \+    \{ return wantarray ? @&#8203;\_ : $\_\[$\#\_\]; \}
&nbsp;   \}

&nbsp;   sub lnoreturn \(\) \{
&nbsp;  @&#8203;@&#8203; \-177\,7 \+177\,7 @&#8203;@&#8203;
&nbsp;           croak "Can't lnoreturn except in ASSIGN context";
&nbsp;       \}
&nbsp;       double\_return\(\);
&nbsp;  \-    return disarm\_temp\(my $undef\);
&nbsp;  \+    \{ return disarm\_temp\(my $undef\); \}
&nbsp;   \}

&nbsp;   \# Some naughty people were relying on these internal methods\.

\-\- Matthew Horsfall \(alh\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From perl5-porters@perl.org

Matthew Horsfall wrote&#8203;:
> What it looks like to me is happening is double\_return pops a stack
> frame\, effectively putting the return calls in the sub above lnoreturn
> and rreturn\.

Yes\, it sets things up in such a way that pp\_return will pop one stack
further than it would have\.

Because pp\_leavesub assumes there is only one stack frame to pop\, it
does not even notice\.

> What I don't know is if Want is violating some rules with regards to
> the Perl internals and doing something it shouldn't\.\.\.

Why else does it exist? :\-\)

> This patch does however fix the issue since the extra scope prevents
> the optimisation from working\.

I think that patch should work\.  Will you do the honours of sub\-
mitting it?

> \(That is\, until we some day make this
> optimisation smarter\)&#8203;:

Applying the optimisation in those cases where the return is nested
inside blocks would result in \*more\* op calls\, because of all the
pp\_leave\(\)s\, etc\.  I think pp\_return/last/goto\, etc\.\, are the fastest
way to exit multiple scopes\.

**Comment by p5pRT** [comment]
### From @wolfsage

On Sun\, Dec 15\, 2013 at 9&#8203;:16 AM\, Father Chrysostomos \<sprout@&#8203;cpan\.org> wrote&#8203;:
> Matthew Horsfall wrote&#8203;:
>> What I don't know is if Want is violating some rules with regards to
>> the Perl internals and doing something it shouldn't\.\.\.
>
> Why else does it exist? :\-\)

Heheh\, fair point\.

>
>> This patch does however fix the issue since the extra scope prevents
>> the optimisation from working\.
>
> I think that patch should work\.  Will you do the honours of sub\-
> mitting it?

Done\. For now I've filed a pull request\, I'll make an RT bug if he
doesn't notice it in a week or so\.

&nbsp;  https://github.com/robinhouston/Want/pull/1

I've taken this the ticket for now\.

Thanks\,

\-\- Matthew Horsfall \(alh\)

**Comment by p5pRT** [comment]
### From @wolfsage

The module author has accepted the patch and pushed a new version&#8203;:

http&#8203;://search\.cpan\.org/~robin/Want\-0\.22/

Closing\.

\-\- Matthew Horsfall \(alh\)

**Comment by p5pRT** [comment]
@wolfsage - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @andk

"Matthew Horsfall via RT" \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:

> The module author has accepted the patch and pushed a new version&#8203;:
>
> http&#8203;://search\.cpan\.org/~robin/Want\-0\.22/
>
> Closing\.

Maybe to be reopened\. Another affected module is JJORE/overload\-eval\-0\.10\.tar\.gz

\-\- 
andreas

**Comment by p5pRT** [comment]
### From @iabyn

On Fri\, May 02\, 2014 at 07&#8203;:43&#8203;:28PM \+0200\, Andreas Koenig wrote&#8203;:
> "Matthew Horsfall via RT" \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:
> 
> > The module author has accepted the patch and pushed a new version&#8203;:
> >
> > http&#8203;://search\.cpan\.org/~robin/Want\-0\.22/
> >
> > Closing\.
> 
> Maybe to be reopened\. Another affected module is JJORE/overload\-eval\-0\.10\.tar\.gz

I see test failures with that module prior to that commit\.

\-\- 
Technology is dominated by two types of people&#8203;: those who understand what
they do not manage\, and those who manage what they do not understand\.

**Comment by p5pRT** [comment]
### From @andk

"Dave Mitchell via RT" \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:

>> > Closing\.
>> 
>> Maybe to be reopened\. Another affected module is JJORE/overload\-eval\-0\.10\.tar\.gz
>
> I see test failures with that module prior to that commit\.

Why so terse? At cpantesters? URL? Commit ID?

\-\- 
andreas

**Comment by p5pRT** [comment]
### From @andk

Andreas Koenig \<andreas\.koenig\.7os6VVqR@&#8203;franz\.ak\.mind\.de> writes&#8203;:

> "Dave Mitchell via RT" \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:
>
>>> > Closing\.
>>> 
>>> Maybe to be reopened\. Another affected module is JJORE/overload\-eval\-0\.10\.tar\.gz
>>
>> I see test failures with that module prior to that commit\.
>
> Why so terse? At cpantesters? URL? Commit ID?

There is no red below the 5\.19\.7 row in the matrix&#8203;:

http&#8203;://matrix\.cpantesters\.org/?dist=overload\-eval%200\.10

\-\- 
andreas

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, May 07\, 2014 at 06&#8203;:55&#8203;:38AM \+0200\, Andreas Koenig wrote&#8203;:
> Andreas Koenig \<andreas\.koenig\.7os6VVqR@&#8203;franz\.ak\.mind\.de> writes&#8203;:
> 
> > "Dave Mitchell via RT" \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:
> >
> >>> > Closing\.
> >>> 
> >>> Maybe to be reopened\. Another affected module is JJORE/overload\-eval\-0\.10\.tar\.gz
> >>
> >> I see test failures with that module prior to that commit\.
> >
> > Why so terse? At cpantesters? URL? Commit ID?
> 
> There is no red below the 5\.19\.7 row in the matrix&#8203;:
> 
> http&#8203;://matrix\.cpantesters\.org/?dist=overload\-eval%200\.10


Sorry\, I meant that I downloaded the module locally\, and tried building
with blead and saw failures\. Then tried with a slightly older 5\.19\.6 perl
binary \(the first release older than the commit in question\) and still saw
failures \(different ones\, but failures still\)\.

But having tried to reproduce it just now\, I'm no longer seeing test
failures on 5\.19\.6\, so maybe I did something wrong earlier\.

I'll look into this further\.

\-\- 
Spock \(or Data\) is fired from his high\-ranking position for not being able
to understand the most basic nuances of about one in three sentences that
anyone says to him\.
&nbsp;    \-\- Things That Never Happen in "Star Trek" \#19

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, May 07\, 2014 at 12&#8203;:17&#8203;:48PM \+0100\, Dave Mitchell wrote&#8203;:
> On Wed\, May 07\, 2014 at 06&#8203;:55&#8203;:38AM \+0200\, Andreas Koenig wrote&#8203;:
> > Andreas Koenig \<andreas\.koenig\.7os6VVqR@&#8203;franz\.ak\.mind\.de> writes&#8203;:
> > >>> Maybe to be reopened\. Another affected module is JJORE/overload\-eval\-0\.10\.tar\.gz

TL;DR&#8203;: its not a regression from 5\.18\.


The commit in question \(which just optimises 'return expr' at the end of a
sub into 'expr'\) turns out to be a red herring\.

After that commit\, you get an assertion failure in t/basic\.t due to a
freed var being accessed\.  The code in question is a perl sub which is
executing the overloaded eval\, by eventually doing the actual eval&#8203;:

&nbsp;    sub wantarray \{
&nbsp;        no overload&#8203;::eval;
&nbsp;        \.\.\.;
&nbsp;        return eval shift;
&nbsp;    \}

The offending commit logically converts the sub into

&nbsp;    sub wantarray \{
&nbsp;        no overload&#8203;::eval;
&nbsp;        \.\.\.;
&nbsp;        eval shift;
&nbsp;    \}

which causes the assertion failure\. \*But\*&#8203;: you get this failure
on perls prior to that commit if you remove the 'return' keyword from the
test file\. So this commit didn't break anything\, fortunately\, it just
flagged an existing issue\.

As to what the underlying issue is\, I'm not exactly sure\. The basic thing
is that if you eval a string\, and the generated code ends with a const op
\- for example\, eval '5;'\, which compiles to const\(5\); leaveeval \- and if the
eval is in void context\, then the const SV is pushed on the stack\, and
leaveval doesn't bother making a copy of it \(since we're in void
context\)\, then the optree is deleted\, including the OP\_CONST\, which frees
the const SV\, leaving a freed SV on the stack&#8203;:

&nbsp;    $ p \-Dst \-e'eval "5;"'

&nbsp;    \.\.\.\.

&nbsp;    \(\-e&#8203;:1\)    entereval
&nbsp;        =>  
&nbsp;    \(\(eval 1\)&#8203;:0\)    nextstate
&nbsp;        =>  
&nbsp;    \(\(eval 1\)&#8203;:1\)    const\(IV\(5\)\)
&nbsp;        =>  IV\(5\)  
&nbsp;    \(\(eval 1\)&#8203;:1\)    leaveeval
&nbsp;        =>  \(FREED\)  
&nbsp;    \(\-e&#8203;:1\)    leave

Normally this is isn't an issue\, but with overload&#8203;::eval allowing you to
create a perl\-level wrapper around the real eval call\, that freed SV is
surviving long enough to cause assertion failures\.

Whether that's an issue for overload&#8203;::eval or perl\, I don't know\.
However\, since I've confirmed that it isn't a 5\.20 regression \(I can get
the same assertion failure in 5\.18\.2 by removing 'return' from t/basic\.t\)\,
I'll let someone else worry about it\.


\-\- 
The Enterprise is involved in a bizarre time\-warp experience which is in
some way unconnected with the Late 20th Century\.
&nbsp;    \-\- Things That Never Happen in "Star Trek" \#14
